### PR TITLE
Keep using the same hyper client

### PR DESCRIPTION
**Pull request by dnlmlr**
- Keep using the same hyper client instead of building a new one for each request
- This allows the client to reuse connections and improves the performance of multiple requests by almost 2x
- The playlist_tracks example takes 38 secs before and 20 secs after the change to enumerate a 180 track playlist
- To avoid carrying the hyper Client generics through the whole project, `ProxyConnector` is always used as the Connector, but disabled when not using a proxy. This should have no real overhead
- The client creation is done lazily to keep the `HttpClient::new` without a `Result` return type

### CODE REVIEW

**Comment by kingosticks** [comment]
I'm sure this is a good change and it is somewhat faster but I'm very surprised by that 2x improvement.

**Comment by dnlmlr** [comment]
Feel free to try it out yourself, I'd be interested in the results. I literally just grabbed a random playlist	(`spotify:playlist:37i9dQZF1DWZeKCadgRdKQ`) from the home screen and ran the playlist example before and after this commit. 
The time measurement was done by simply taking the time before the loop start and printing the elapsed time after the loop. I'll add the actual code when I'm on my desktop PC.

**Comment by dnlmlr** [comment]
The exact code I used for testing was this modification to the `examples/playlist_tracks.rs`:
```diff
     let plist = Playlist::get(&session, &plist_uri).await.unwrap();
-    println!("{:?}", plist);
+    // println!("{:?}", plist);
+    let t1 = std::time::Instant::now();
     for track_id in plist.tracks() {
         let plist_track = Track::get(&session, track_id).await.unwrap();
         println!("track: {} ", plist_track.name);
     }
+    println!("{}", t1.elapsed().as_secs_f64());
```
This is an extremely basic test and includes the print time and and iterator creation within the measured time, but running it multiple times gives me pretty stable results (less than +-1 sec).

I just quickly redit the test and recorded the result I got (times in seconds, lower is better):

| Test | Avg. | Run 1 | Run 2 | Run 3 | Speed vs New client / request |
| ---- | ---- | ------- | ------ | ----- | ------------------------------ |
| New client / request (before this PR) | **37.45** | 37.62 |  37.61 |  37.11 | 1x |
| Keeping client (this PR) | **20.02** | 19.66 | 19.80 | 20.60 | 1.87x |
| Mercury instead of HTTP | **10.46** | 10.39 | 10.31 | 10.69 | 3.58x |

Mercury is also included just for reference. For that test I changed the `metadata::Track` back to mercury like this:
```diff
     async fn request(session: &Session, track_id: &SpotifyId) -> RequestResult {
-        session.spclient().get_track_metadata(track_id).await
+        // session.spclient().get_track_metadata(track_id).await
+        let uri = format!("hm://metadata/3/track/{}", track_id.to_base16()?);
+        <Self as crate::request::MercuryRequest>::request(session, &uri).await
     }
```

PS: I'm not trying to preassure or even suggest reverting back to mercury for performance reasons. I'm simply trying to get the best performance with what is being used and having all the reference data is really useful context for that.

**Comment by kingosticks** [comment]
Thank you for the code, when I'm back from holidays I'll give it a go but only to satisfy my curiosity. This is a sensible improvement either way. Good work.

**Comment by kingosticks** [comment]
Actually, relatedly, does anyone know if there's any rate limiting with this? I keep hitting that with my use of the Web API and keen to avoid it.

**Comment by roderickvd** [comment]
Great work again. Definitely an improvement as it shows. As I’m away from my development rig I can’t test and wonder about a few things:

- What do you think happens under the `hyper` hood? Is the connection kept alive?

- If so how does this go when the client is used for different hosts? Does it transparently keep or close and reopen connections depending on whether consecutive requests are for the same host or not?

- Does this work well not only in the case of metadata (which is all going to the `spclient` host) but also with a full playback session, which makes many different requests to different hosts (also the CDN, resolver, token server)?

- Would it make sense to have and keep multiple clients, dedicated to `spclient`, another to the CDN, and so on?

This train of thought hinges entirely on the thought of keep alive, is that what’s at play here or something else? I would need to pop that `hyper` hood.

**Comment by dnlmlr** [comment]
I'm in no way an expert for `hyper`, but I've used it a few times in the past and I'll try to address your points:

- Under the hood, the client tries to keep the connections open. It keeps a pool of idle connection connections that will be reuesed for requests to the same host. These connections will timeout after 90 seconds by default and there can be multiple idle connections for the same host.
- When connecting to any host, it will check if there are currently open connections in the pool and if not it will create a new connection which will then be added to the pool.
- Performance improvements will exist in any case where there are multiple requests to the same host in a relatively small timeframe (idle connections time out after 90 secs default). But keep in mind that the only improved part is the connection establishment, so for requests like playback with few large requests there will probably not be a big improvement. On the other hand there should not be a downside either
- As before the connections are stored in a pool and the pool can actually even be shared. I believe using the same pool (or even clients) for all requests is the intended way to use the client. Having separate pools would restrict each client to only connections from its own pool. 
  > Client is cheap to clone and cloning is the recommended way to share a Client. The underlying connection pool will be reused.

To summarize: I think using either the same client (or at clones of the same client to keep a shared connection pool) is the intended way to use the hyper client. It will try to use idle connections from the pool if available, or make new connections and try to keep them idle, adding them to the pool. By default connections will be dropped after 90 seconds.

https://docs.rs/hyper/latest/hyper/client/index.html#client

**Comment by roderickvd** [comment]
Perfect. From the `hyper` documentation it seems that HTTP/2 keep alive is disabled by default while I know that at least the CDN uses HTTP/2. Any reason not to enable?

**Comment by kingosticks** [comment]
I was under the impression the keep alive request header doesn't make sense for http2, multiple requests are multiplexed over the same connection so it's kept open by default.

**Comment by roderickvd** [comment]
Indeed it’s not a header but in the case of HTTP/2 you need to send a ping for the connection to remain open.

**Comment by kingosticks** [comment]
Apologies, you are right!

**Comment by roderickvd** [comment]
> Actually, relatedly, does anyone know if there's any rate limiting with this? I keep hitting that with my use of the Web API and keen to avoid it.

There probably is but I don’t know the limits. It’s on my list to add a proper rate limiter in the HTTP client. Just a matter of selecting a good crate and adding a couple of lines, probably.

**Comment by dnlmlr** [comment]
> Perfect. From the `hyper` documentation it seems that HTTP/2 keep alive (ping) is disabled by default while I know that at least the CDN uses HTTP/2. Any reason not to enable?

I don't think the http/2 ping is strictly necessary to keep open the connection with h2. Afaik the default behavior for h2 clients is to keep the connection open until it is certain that no more requests will be sent. The server should also try to keep it open as long as possible, but can close at any time for example due to resource constraints. This is roughly the same behavior that you have with http/1 when using the keep-alive header. 

That being said, I don't see a reason to not enable the ping frames either. I assume it could still help to avoid connection termination in some circumstances

**Comment by roderickvd** [comment]
Alright. Let’s get this in and we can see about the ping later. Only use case I can think of is when skipping in or between songs after so many seconds of listening.
